src/librustc_mir/interpret/terminator.rs

```diff
@@ -196,6 +196,9 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M>
                 // that will take care to make it UB to leave the range, just
                 // like for transmute).
                 caller.value == callee.value,
+            (layout::Abi::ScalarPair(ref caller1, ref caller2),
+             layout::Abi::ScalarPair(ref callee1, ref callee2)) =>
+                caller1.value == callee1.value && caller2.value == callee2.value,
```

Comment:
Are you not using `FnType` yet :P? Anyway this seems... fine.
*But only* for the `Rust` ABI so you might want to add that as a condition, if it's not already.
(`ScalarPair` is considered an aggregate and follows more complex rules, in non-`Rust` ABIs)

Reply:
`Abi` does not encode everything, no. C ABIs can depend on details of the C type definition, which is annoying but `FnType` contains all of that information.

> Nope, you didn't yet convince me that it would make anything simpler.^^

It would make things *correct*, which seems more important than simplicity. It's not like that hard to use, just have to move some code from `rustc_codegen_ssa` and then compare `ArgType` pairwise.

(assuming you don't want to emulate the runtime behavior of mismatched ABIs :P)

Reply:
> C ABIs can depend on details of the C type definition, which is annoying but FnType contains all of that information.

Ouch. Okay, I will restrict type mismatches to Rust Abi functions.

> assuming you don't want to emulate the runtime behavior of mismatched ABIs :P

You assume correctly. :P

> It's not like that hard to use, just have to move some code from rustc_codegen_ssa and then compare ArgType pairwise.

Would you mentor someone? I could open an E-mentor issue.

Reply:
> Would you mentor someone? I could open an E-mentor issue.

Sure, always happy to mentor compiler cleanups!

The main issue to moving that stuff around is `pointee_info_at` (it would have to be moved to `rustc::ty::layout`, as an additional method in `TyLayoutMethods`):
https://github.com/rust-lang/rust/blob/dae6c9364108847393075b57a17fd7eaf3f37600/src/librustc_codegen_llvm/abi.rs#L487

Everything else is mostly relying on `rustc_target` doing the heavy lifting, already.

Reply:
> But only for the Rust ABI so you might want to add that as a condition

Done.